### DISCUSSION TITLE
Databases URL in a separate Files (Instead of .env)

### DISCUSSION THREAD

**Discussion by alan345** [prisma/prisma#8787] [Q&A]
## My Problem 
I have many databases, (hence many Prisma schema)  in many environments. I need to centralize all the databases access into 1 config file. Using `.env` files is not a solution as it is too messy and hard to maintain.

### env1
- service1
  - schema.prisma
  - .env (DATABASE_URL = URL1)
- service2
  - schema.prisma
  - .env (DATABASE_URL = URL2)
- service3
  - schema.prisma
  - .env (DATABASE_URL = URL3) 

### env2
- service1
  - schema.prisma
  - .env (DATABASE_URL = URL4)
- service2
  - schema.prisma
  - .env (DATABASE_URL = URL5)
- service3
  - schema.prisma
  - .env (DATABASE_URL = URL6) 

## Ideal Solution
The best solution would be to store all the Databases URLs into 1 file called `config.ts`
and inside each `schema.prisma` I will call this config file.

```
// schema.prisma

import config from '../config/config'
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mysql"
  url      = config.URL1
}
```

![image](https://user-images.githubusercontent.com/15246526/129600589-3e5c2eae-fbd7-4e04-95c8-d10724a65529.png)

Remark:
```
const prisma = new PrismaClient({
  datasources: {
    db: {
      url: config.URL_DB_NN,
    },
  },
})
```
Using this in the code doesn't work if the app if I run `npx prisma migrate deploy`

What would be the best workaround?

Related to: https://github.com/prisma/prisma/issues/2443
documentation: https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#programmatically-override-a-datasource-url

**Comment by janpio** [comment]
I would use `datasource.db.url` (your last code snippet) for Prisma Client usage, and use some other script to get the correct value into an actual ENV var from the schema file for Prisma CLI usage (or even use that script for both to keep the code clean and simple).

**Reply by alan345** [reply]
Thanks @janpio . It is what we are doing now (script to populate .env files). But it is not very clean and convenient

**Reply by janpio** [reply]
I would skip the `.env` files indeed, and populate the environment variables directly indeed.

**Reply by carlosyan1807** [reply]
Yes, not everyone uses `dotenv`, and in some closed projects, environment variables are directly written into config files, such as `config.dev.ts` or `config.prod.json`. These code repositories are privately deployed, so No need to consider security issues. Now the way has to maintain two or more environment configuration, which is inconvenient. if `schema.prisma` can load these config will be a very cool thing.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
